#  Battle of the Neighbourhoods - Toronto

This notebook contains Questions 1, 2 & 3 of the Assignment. They have been segregated by Section headers

## Question 1


In [1]:
import pandas as pd
import numpy as np
import requests

In [10]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
wiki_url = requests.get(url)
wiki_url

<Response [200]>

In [11]:
wiki_data = pd.read_html(wiki_url.text)
wiki_data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                          Neighbourhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

In [12]:
len(wiki_data), type(wiki_data)

(3, list)

We need the first table alone, so dropping the other tables

In [13]:
wiki_data = wiki_data[0]
wiki_data

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

In [14]:
df = wiki_data[wiki_data["Borough"] != "Not assigned"]
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Grouping the records based on Postal Code

In [15]:
df = df.groupby(['Postal Code']).head()
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [16]:
df.Neighbourhood.str.count("Not assigned").sum()

0

In [17]:
df = df.reset_index()
df

index Postal Code           Borough  \
0        2         M3A        North York   
1        3         M4A        North York   
2        4         M5A  Downtown Toronto   
3        5         M6A        North York   
4        6         M7A  Downtown Toronto   
..     ...         ...               ...   
98     160         M8X         Etobicoke   
99     165         M4Y  Downtown Toronto   
100    168         M7Y      East Toronto   
101    169         M8Y         Etobicoke   
102    178         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 4 columns]

In [18]:
df.drop(['index'], axis = 'columns', inplace = True)
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [19]:
df.shape

(103, 3)

Answer to Question 1: We have 103 rows and 3 columns

## Question 2

In [20]:
pip install geocoder

     |████████████████████████████████| 98 kB 9.4 MB/s  eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [21]:
import geocoder # import geocoder

In [22]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [23]:
print("The shape of our wiki data is: ", df.shape)
print("the shape of our csv data is: ", data.shape)

The shape of our wiki data is:  (103, 3)
the shape of our csv data is:  (103, 3)


In [24]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

In [25]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [26]:
combined_data = df.join(data.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [27]:
combined_data.shape

(103, 5)

Solution: We get 103 rows as expected when we do a inner join, so we have good data.

## Question 3

Drawing inspiration from the previous lab where we cluster the neighbourhood of NYC, We cluster Toronto based on the similarities of the venues categories using Kmeans clustering and Foursquare API.

In [28]:
from geopy.geocoders import Nominatim

In [29]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


In [31]:
!pip install folium

     |████████████████████████████████| 93 kB 3.7 MB/s  eta 0:00:01


In [32]:
import folium

In [33]:
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

In [34]:
CLIENT_ID = '2JE15UX221MJBOHVPUVJIUHNCRIZUAIBRHPTHMH41AS31ND2' 
CLIENT_SECRET = 'B4TVHPDUMYNGGV4M11PVNGITEQZIOXMOWCJ2OP04ZW1UHREL'
VERSION = '20180604' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 2JE15UX221MJBOHVPUVJIUHNCRIZUAIBRHPTHMH41AS31ND2
CLIENT_SECRET:B4TVHPDUMYNGGV4M11PVNGITEQZIOXMOWCJ2OP04ZW1UHREL


Next, we create a function to get all the venue categories in Toronto

In [36]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

In [37]:
venues_in_toronto = getNearbyVenues(combined_data['Neighbourhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [38]:
venues_in_toronto.shape

(1333, 5)

In [39]:
venues_in_toronto.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2         Parkwoods               43.753259               -79.329656   
3         Parkwoods               43.753259               -79.329656   
4  Victoria Village               43.725882               -79.315572   

                    Venue        Venue Category  
0         Brookbanks Park                  Park  
1         Brookbanks Pool                  Pool  
2                     KFC  Fast Food Restaurant  
3           Variety Store     Food & Drink Shop  
4  Victoria Village Arena          Hockey Arena

In [40]:
venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                             Parkwoods   
3                                             Parkwoods   
4                                      Victoria Village   
...                                                 ...   
1318  Mimico NW, The Queensway West, South of Bloor,...   
1319  Mimico NW, The Queensway West, South of Bloor,...   
1320  Mimico NW, The Queensway West, South of Bloor,...   
1321  Mimico NW, The Queensway West, South of Bloor,...   
1322  Mimico NW, The Queensway West, South of Bloor,...   

      Neighbourhood Latitude  Neighbourhood Longitude                   Venue  \
0                  43.753259               -79.329656         Brookbanks Park   
1                  43.753259               -79.329656         Brookbanks Pool   
2                  43.753259               -79.329656                     KFC   
3                  43.753259               -79.329656           Variety Store   
4                  43.725882               -79.315572  Victoria Village Arena   
...                      ...                      ...                     ...   
1318               43.628841               -79.520999              Wingporium   
1319               43.628841               -79.520999        South St. Burger   
1320               43.628841               -79.520999               Dollarama   
1321               43.628841               -79.520999          Healthy Planet   
1322               43.628841               -79.520999    Artisano Bakery Café   

            Venue Category  
0                     Park  
1                     Pool  
2     Fast Food Restaurant  
3        Food & Drink Shop  
4             Hockey Arena  
...                    ...  
1318           Wings Joint  
1319          Burger Joint  
1320        Discount Store  
1321       Supplement Shop  
1322                Bakery  

[416 rows x 5 columns]

So there are 405 records for each neighbourhood.

Checking for the maximum venue categories

In [41]:
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                          
Accessories Store                    Lawrence Manor, Lawrence Heights   
Adult Boutique                                   Church and Wellesley   
Airport                                                     Downsview   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Gate        CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Warehouse Store                                      Thorncliffe Park   
Wine Bar                                     Little Portugal, Trinity   
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                                     Caledonia-Fairbanks   
Yoga Studio                            University of Toronto, Harbord   

                    Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                        
Accessories Store                43.718518               -79.464763   
Adult Boutique                   43.665860               -79.383160   
Airport                          43.737473               -79.394420   
Airport Food Court               43.628947               -79.394420   
Airport Gate                     43.628947               -79.394420   
...                                    ...                      ...   
Warehouse Store                  43.705369               -79.349372   
Wine Bar                         43.653206               -79.400049   
Wings Joint                      43.628841               -79.520999   
Women's Store                    43.689026               -79.453512   
Yoga Studio                      43.715383               -79.321558   

                                              Venue  
Venue Category                                       
Accessories Store               Ardene Shoes Outlet  
Adult Boutique                            Seduction  
Airport             Toronto Downsview Airport (YZD)  
Airport Food Court                Billy Bishop Café  
Airport Gate                                 Gate 8  
...                                             ...  
Warehouse Store                              Costco  
Wine Bar                            Paris Paris Bar  
Wings Joint                              Wingporium  
Women's Store                         Maximum Woman  
Yoga Studio                               YogaSpace  

[238 rows x 4 columns]

## One Hot encoding the venue Categories

In [42]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Adult Boutique  Airport  Airport Food Court  \
0                     0               0        0                   0   
1                     0               0        0                   0   
2                     0               0        0                   0   
3                     0               0        0                   0   
4                     0               0        0                   0   
...                 ...             ...      ...                 ...   
1328                  0               0        0                   0   
1329                  0               0        0                   0   
1330                  0               0        0                   0   
1331                  0               0        0                   0   
1332                  0               0        0                   0   

      Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0                0               0                0                 0   
1                0               0                0                 0   
2                0               0                0                 0   
3                0               0                0                 0   
4                0               0                0                 0   
...            ...             ...              ...               ...   
1328             0               0                0                 0   
1329             0               0                0                 0   
1330             0               0                0                 0   
1331             0               0                0                 0   
1332             0               0                0                 0   

      American Restaurant  Antique Shop  ...  Train Station  Truck Stop  \
0                       0             0  ...              0           0   
1                       0             0  ...              0           0   
2                       0             0  ...              0           0   
3                       0             0  ...              0           0   
4                       0             0  ...              0           0   
...                   ...           ...  ...            ...         ...   
1328                    0             0  ...              0           0   
1329                    0             0  ...              0           0   
1330                    0             0  ...              0           0   
1331                    0             0  ...              0           0   
1332                    0             0  ...              0           0   

      Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                                 0                 0                      0   
1                                 0                 0                      0   
2                                 0                 0                      0   
3                                 0                 0                      0   
4                                 0                 0                      0   
...                             ...               ...                    ...   
1328                              0                 0                      0   
1329                              0                 0                      0   
1330                              0                 0                      0   
1331                              0                 0                      0   
1332                              0                 0                      0   

      Warehouse Store  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                   0         0            0              0            0  
1                   0         0            0              0            0  
2                   0         0            0              0            0  
3                   0         0            0              0            0  
4                   0      

Adding the neighbourhood to the encoded dataframe

In [43]:
toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Neighbourhood  Accessories Store  Adult Boutique  Airport  \
0         Parkwoods                  0               0        0   
1         Parkwoods                  0               0        0   
2         Parkwoods                  0               0        0   
3         Parkwoods                  0               0        0   
4  Victoria Village                  0               0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  ...  Train Station  Truck Stop  \
0                 0                    0  ...              0           0   
1                 0                    0  ...              0           0   
2                 0                    0  ...              0           0   
3                 0                    0  ...              0           0   
4                 0                    0  ...              0           0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Warehouse Store  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                0         0            0              0            0  
1                0         0            0              0            0  
2                0         0            0              0            0  
3                0         0            0              0            0  
4                0         0            0              0            0  

[5 rows x 239 columns]

We will group the Neighbourhoods, calculate the mean venue categories in each Neighbourhood

In [44]:
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Adult Boutique  Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0             0.0      0.0                 0.0           0.0             0.0   
1             0.0      0.0                 0.0           0.0             0.0   
2             0.0      0.0                 0.0           0.0             0.0   
3             0.0      0.0                 0.0           0.0             0.0   
4             0.0      0.0                 0.0           0.0             0.0   

   Airport Service  Airport Terminal  American Restaurant  ...  Train Station  \
0              0.0               0.0             0.000000  ...            0.0   
1              0.0               0.0             0.000000  ...            0.0   
2              0.0               0.0             0.000000  ...            0.0   
3              0.0               0.0             0.000000  ...            0.0   
4              0.0               0.0             0.043478  ...            0.0   

   Truck Stop  Vegetarian / Vegan Restaurant  Video Game Store  \
0         0.0                            0.0               0.0   
1         0.0                            0.0               0.0   
2         0.0                            0.0               0.0   
3         0.0                            0.0               0.0   
4         0.0                            0.0               0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                    0.0              0.0       0.0          0.0   
1                    0.0              0.0       0.0          0.0   
2                    0.0              0.0       0.0          0.0   
3                    0.0              0.0       0.0          0.0   
4                    0.0              0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 239 columns]

Let's make a function to get the top most common venue categories

In [45]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [46]:
import numpy as np

There are way too many venue categories, we can take the top 10 to cluster the neighbourhoods

In [47]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                  Bank   
3                                  Bayview Village    Chinese Restaurant   
4                Bedford Park, Lawrence Manor East       Thai Restaurant   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0  Latin American Restaurant          Skating Rink        Breakfast Spot   
1                Coffee Shop                   Gym        Sandwich Place   
2                Coffee Shop      Sushi Restaurant           Supermarket   
3                       Bank   Japanese Restaurant                  Café   
4                Coffee Shop        Sandwich Place    Italian Restaurant   

  5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
0      Department Store     Electronics Store  Eastern European Restaurant   
1                   Pub                 Diner             Department Store   
2        Ice Cream Shop            Restaurant    Middle Eastern Restaurant   
3           Yoga Studio      Department Store            Electronics Store   
4      Greek Restaurant             Juice Bar                   Restaurant   

         8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                    Drugstore            Donut Shop                Dog Run  
1                 Dessert Shop    Dim Sum Restaurant    Distribution Center  
2                Deli / Bodega     Convenience Store            Bridal Shop  
3  Eastern European Restaurant             Drugstore             Donut Shop  
4            Indian Restaurant      Sushi Restaurant                    Pub

Let's make the model to cluster our Neighbourhoods

In [48]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [49]:
# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(n_clusters=5, random_state=0)

In [50]:
kmeans.labels_[0:100]

array([3, 0, 3, 3, 3, 3, 3, 3, 3, 3, 2, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3,
       3, 3, 3, 3, 2, 3, 3, 3, 3, 3, 3, 1, 3, 3, 3, 3, 0, 3, 3, 3, 3, 3,
       1, 3, 1, 3, 3, 3, 1, 3, 3, 3, 1, 3, 3, 3, 3, 3, 3, 1, 3, 3, 3, 2,
       3, 1, 3, 3, 4, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 2, 3, 3, 3, 0, 0, 2,
       1, 2, 3, 0, 3, 3, 2], dtype=int32)

In [51]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Join toronto_grouped with combined_data on neighbourhood to add latitude & longitude for each neighborhood to prepare it for plotting

In [52]:
toronto_merged = combined_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656             1.0                  Park   
1  43.725882 -79.315572             0.0           Coffee Shop   
2  43.654260 -79.360636             3.0           Coffee Shop   
3  43.718518 -79.464763             3.0        Clothing Store   
4  43.662301 -79.389494             3.0           Coffee Shop   

    2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
0                    Pool   Fast Food Restaurant     Food & Drink Shop   
1            Hockey Arena  Portuguese Restaurant           Pizza Place   
2                    Park                 Bakery        Breakfast Spot   
3  Furniture / Home Store      Accessories Store    Athletics & Sports   
4        Sushi Restaurant            Yoga Studio        Sandwich Place   

         5th Most Common Venue 6th Most Common Venue  \
0                  Yoga Studio         Deli / Bodega   
1  Eastern European Restaurant             Drugstore   
2                         Café     French Restaurant   
3                  Event Space    Miscellaneous Shop   
4         Fast Food Restaurant    Italian Restaurant   

         7th Most Common Venue 8th Most Common Venue  9th Most Common Venue  \
0  Eastern European Restaurant             Drugstore             Donut Shop   
1                   Donut Shop               Dog Run    Distribution Center   
2          Distribution Center                   Spa            Event Space   
3                     Boutique           Coffee Shop  Vietnamese Restaurant   
4          Japanese Restaurant              Beer Bar          Smoothie Shop   

  10th Most Common Venue  
0                Dog Run  
1            Curling Ice  
2           Dessert Shop  
3        Airport Service  
4    Distribution Center

In [53]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

In [54]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [55]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

Let's verify each of our clusters

Cluster 1

In [56]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
1   North York             0.0           Coffee Shop          Hockey Arena   
50  North York             0.0           Pizza Place          Intersection   
70   Etobicoke             0.0           Pizza Place           Coffee Shop   
72  North York             0.0              Pharmacy           Coffee Shop   
93   Etobicoke             0.0           Pizza Place           Coffee Shop   

    3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
1   Portuguese Restaurant           Pizza Place  Eastern European Restaurant   
50   Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   
70     Chinese Restaurant        Discount Store               Sandwich Place   
72            Pizza Place         Grocery Store                    Gift Shop   
93                    Gym        Sandwich Place                          Pub   

   6th Most Common Venue      7th Most Common Venue 8th Most Common Venue  \
1              Drugstore                 Donut Shop               Dog Run   
50             Drugstore                 Donut Shop               Dog Run   
70          Intersection  Middle Eastern Restaurant         Deli / Bodega   
72           Curling Ice                  Drugstore            Donut Shop   
93                 Diner           Department Store          Dessert Shop   

   9th Most Common Venue 10th Most Common Venue  
1    Distribution Center            Curling Ice  
50   Distribution Center         Discount Store  
70             Drugstore             Donut Shop  
72               Dog Run    Distribution Center  
93    Dim Sum Restaurant    Distribution Center

Cluster 2

In [57]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels      1st Most Common Venue  \
0        North York             1.0                       Park   
12      Scarborough             1.0                        Bar   
44      Scarborough             1.0                     Bakery   
49       North York             1.0                       Park   
61  Central Toronto             1.0                       Park   
71      Scarborough             1.0  Middle Eastern Restaurant   
77        Etobicoke             1.0          Mobile Phone Shop   
85      Scarborough             1.0                       Park   

   2nd Most Common Venue       3rd Most Common Venue 4th Most Common Venue  \
0                   Pool        Fast Food Restaurant     Food & Drink Shop   
12           Yoga Studio                 Event Space           Escape Room   
44              Bus Line                        Park          Soccer Field   
49      Basketball Court  Construction & Landscaping                Bakery   
61           Swim School            Business Service              Bus Line   
71         Shopping Mall                      Bakery        Sandwich Place   
77                  Park              Sandwich Place              Bus Line   
85            Playground                Intersection           Yoga Studio   

   5th Most Common Venue        6th Most Common Venue  \
0            Yoga Studio                Deli / Bodega   
12     Electronics Store  Eastern European Restaurant   
44         Metro Station                  Bus Station   
49           Yoga Studio                 Dessert Shop   
61         Deli / Bodega            Electronics Store   
71           Auto Garage                  Escape Room   
77          Dance Studio                    Drugstore   
85          Dance Studio  Eastern European Restaurant   

          7th Most Common Venue 8th Most Common Venue  \
0   Eastern European Restaurant             Drugstore   
12                    Drugstore            Donut Shop   
44               Ice Cream Shop     Electronics Store   
49                  Escape Room     Electronics Store   
61  Eastern European Restaurant             Drugstore   
71  Eastern European Restaurant             Drugstore   
77                   Donut Shop               Dog Run   
85                    Drugstore            Donut Shop   

          9th Most Common Venue 10th Most Common Venue  
0                    Donut Shop                Dog Run  
12                      Dog Run    Distribution Center  
44  Eastern European Restaurant              Drugstore  
49  Eastern European Restaurant              Drugstore  
61                   Donut Shop                Dog Run  
71                   Donut Shop                Dog Run  
77          Distribution Center         Discount Store  
85                      Dog Run    Distribution Center

Cluster 3

In [58]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
21              York             2.0                  Park   
35         East York             2.0                  Park   
52        North York             2.0                  Park   
64              York             2.0                  Park   
66        North York             2.0                  Park   
91  Downtown Toronto             2.0                  Park   
98         Etobicoke             2.0                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
21         Women's Store                  Pool          Dance Studio   
35     Convenience Store           Yoga Studio         Deli / Bodega   
52           Yoga Studio          Dance Studio     Electronics Store   
64     Convenience Store           Yoga Studio         Deli / Bodega   
66     Convenience Store           Yoga Studio         Deli / Bodega   
91                 Trail            Playground           Yoga Studio   
98                 River           Yoga Studio          Dance Studio   

          5th Most Common Venue        6th Most Common Venue  \
21            Electronics Store  Eastern European Restaurant   
35            Electronics Store  Eastern European Restaurant   
52  Eastern European Restaurant                    Drugstore   
64            Electronics Store  Eastern European Restaurant   
66            Electronics Store  Eastern European Restaurant   
91                Deli / Bodega             Department Store   
98            Electronics Store  Eastern European Restaurant   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
21             Drugstore            Donut Shop               Dog Run   
35             Drugstore            Donut Shop               Dog Run   
52            Donut Shop               Dog Run   Distribution Center   
64             Drugstore            Donut Shop               Dog Run   
66             Drugstore            Donut Shop               Dog Run   
91          Dessert Shop    Dim Sum Restaurant                 Diner   
98             Drugstore            Donut Shop               Dog Run   

   10th Most Common Venue  
21    Distribution Center  
35    Distribution Center  
52         Discount Store  
64    Distribution Center  
66    Distribution Center  
91    Distribution Center  
98    Distribution Center